Products/ZenModel/MaintenanceWindow.py

```diff
@@ -645,6 +647,31 @@ def _setProdState(devices_batch):
                 else:
                     device.setProdState(minProdState, maintWindowChange=True)
 
+        def retrySingleDevices(devices_batch):
+            log.warn("Retrying devices individually")
+            for dev in devices_batch:
+                try:
+                    processFunc([dev])
+                except Exception:
+                    log.exception("The production stage change for %s raised the exception." % str(dev))
```

Comment:
The logging library will perform string interpolation for you when the message is written to the log.  This means you can write the logging call like this instead:

```
    log.exception(
        "The production stage change for %s raised the exception.", dev,
    )
```
There's also no need to call `str(dev)` because the `%s` placeholder in the format string will call `str()` on the argument for you.  FYI, the `%r` placeholder will call `repr()` on the argument.

Reply:
fixed